pkg/steps/multi_stage.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/openshift/ci-tools/pkg/junit"
+)
+
+const (
+	multiStageTestLabel = "ci.openshift.io/multi-stage-test"
+)
+
+type multiStageTestStep struct {
+	name            string
+	config          *api.ReleaseBuildConfiguration
+	podClient       PodClient
+	jobSpec         *api.JobSpec
+	pre, test, post []api.TestStep
+	subTests        []*junit.TestCase
+}
+
+func MultiStageTestStep(
+	testConfig api.TestStepConfiguration,
+	config *api.ReleaseBuildConfiguration,
+	podClient PodClient,
+	jobSpec *api.JobSpec,
+) (api.Step, error) {
+	return newMultiStageTestStep(testConfig, config, podClient, jobSpec), nil
+}
+
+func newMultiStageTestStep(
+	testConfig api.TestStepConfiguration,
+	config *api.ReleaseBuildConfiguration,
+	podClient PodClient,
+	jobSpec *api.JobSpec,
+) *multiStageTestStep {
+	return &multiStageTestStep{
+		name:      testConfig.As,
+		config:    config,
+		podClient: podClient,
+		jobSpec:   jobSpec,
+		pre:       testConfig.MultiStageTestConfiguration.Pre,
+		test:      testConfig.MultiStageTestConfiguration.Test,
+		post:      testConfig.MultiStageTestConfiguration.Post,
+	}
+}
+
+func (s *multiStageTestStep) Inputs(ctx context.Context, dry bool) (api.InputDefinition, error) {
+	return nil, nil
+}
+
+func (s *multiStageTestStep) Run(ctx context.Context, dry bool) error {
+	if dry {
+		return nil
+	}
+	var errs []error
+	if err := s.runSteps(ctx, s.pre, true); err != nil {
+		errs = append(errs, fmt.Errorf("%q pre steps failed: %v", s.name, err))
+	} else if err := s.runSteps(ctx, s.test, true); err != nil {
+		errs = append(errs, fmt.Errorf("%q test steps failed: %v", s.name, err))
+	}
+	if err := s.runSteps(ctx, s.post, false); err != nil {
+		errs = append(errs, fmt.Errorf("%q post steps failed: %v", s.name, err))
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func (s *multiStageTestStep) Done() (bool, error) { return false, nil }
+func (s *multiStageTestStep) Name() string        { return s.name }
+func (s *multiStageTestStep) Description() string {
+	return fmt.Sprintf("Run test %s", s.name)
+}
+
+func (s *multiStageTestStep) Requires() (ret []api.StepLink) {
+	var needsImages, needsRelease bool
+	for _, step := range append(append(s.pre, s.test...), s.post...) {
+		if s.config.IsPipelineImage(step.Image) {
+			ret = append(ret, api.InternalImageLink(api.PipelineImageStreamTagReference(step.Image)))
+		} else if s.config.BuildsImage(step.Image) {
+			needsImages = true
+		} else {
+			needsRelease = true
+		}
+	}
+	if needsImages {
+		ret = append(ret, api.ImagesReadyLink())
+	}
+	if needsRelease {
+		ret = append(ret, api.ReleaseImagesLink())
+	}
+	return
+}
+
+func (s *multiStageTestStep) Creates() []api.StepLink { return nil }
+func (s *multiStageTestStep) Provides() (api.ParameterMap, api.StepLink) {
+	return nil, nil
+}
+func (s *multiStageTestStep) SubTests() []*junit.TestCase { return s.subTests }
+
+func (s multiStageTestStep) runSteps(ctx context.Context, steps []api.TestStep, shortCircuit bool) error {
```

Comment:
Hmm... is the non-pointer receiver here intentional?

Reply:
Yes. Is there a problem?

Reply:
Well, that's what I asked.

Using a value receiver was based on the (reasonable but not optimal) assumption that it is what methods that don't mutate the object should use. After reading some golang [nonsense](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type), I suppose the size of the object is the contention here? Changed.

Reply:
If we don't mutate the object, we don't need to pass the pointer.